src/hotspot/share/interpreter/interpreterRuntime.cpp

```diff
@@ -733,20 +733,23 @@ JRT_ENTRY_NO_ASYNC(void, InterpreterRuntime::monitorenter(JavaThread* thread, Ba
 JRT_END
 
 
-//%note monitor_1
-JRT_ENTRY_NO_ASYNC(void, InterpreterRuntime::monitorexit(JavaThread* thread, BasicObjectLock* elem))
+JRT_LEAF(void, InterpreterRuntime::monitorexit(JavaThread* thread, BasicObjectLock* elem))
 #ifdef ASSERT
   thread->last_frame().interpreter_frame_verify_monitor(elem);
 #endif
-  Handle h_obj(thread, elem->obj());
-  assert(Universe::heap()->is_in_or_null(h_obj()),
-         "must be NULL or an object");
-  if (elem == NULL || h_obj()->is_unlocked()) {
-    THROW(vmSymbols::java_lang_IllegalMonitorStateException());
+  oop obj = elem->obj();
+  assert(Universe::heap()->is_in(obj), "must an object");
+  // The object could become unlocked through a JNI call, which we have no other checks for.
+  // Give a message if CheckJNICalls but ignore.
```

Comment:
That comment is a bit misleading. Perhaps:
// Give a fatal message if CheckJNICalls. Otherwise we ignore it.